File: ds-grid/src/distributed/systems/grid/simulation/Simulation.java
package distributed.systems.grid.simulation;

import java.util.ArrayList;
import java.util.List;

import distributed.systems.grid.model.ResourceManager;
import distributed.systems.grid.model.Scheduler;
import distributed.systems.grid.model.Socket;
import distributed.systems.grid.model.User;
import distributed.systems.grid.model.Worker;

/**
 * This class should initialize every other class and ultimately start the
 * simulation
 */
public class Simulation {
	
	private SimulationContext context;
	
	public Simulation(SimulationContext context, int schedulerCount, int clusterCount, int workerCount) {
		this.context = context.register(this);
		
		List<Socket> schedulerSockets = new ArrayList<Socket>();
		List<Socket> rmSockets = new ArrayList<Socket>();

		List<ResourceManager> resourceManagers = new ArrayList<ResourceManager>();
		for (int i = 0; i < clusterCount; i++) {			
			ResourceManager newResourceManager = new ResourceManager(this.context);
			resourceManagers.add(newResourceManager);
			Socket rmSocket = newResourceManager.getSocket();
			rmSockets.add(rmSocket);
			
			// CREATE WORKERS AFTER RM :)
			List<Socket> workerSockets = new ArrayList<Socket>();
			ArrayList<Worker> workers = new ArrayList<Worker>();
			for (int j = 0; j < workerCount; j++) {
				Worker worker = new Worker(this.context);
				workers.add(worker);
				workerSockets.add(worker.getSocket());
			}
			
			newResourceManager.setWorkerSockets(workerSockets);
		}

		List<Scheduler> schedulers = new ArrayList<Scheduler>();
		for (int i = 0; i < schedulerCount; i++) {
			Scheduler newScheduler = new Scheduler(this.context, rmSockets);
			schedulers.add(newScheduler);
			schedulerSockets.add(newScheduler.getSocket());
		}
		
		/* let the schedulers know about eachother */
		for (int i = 0; i < schedulers.size(); i++) {
			schedulers.get(i).setSchedulers(schedulerSockets);
		}

		System.out.printf(">> Done with initializing all nodes\n");
		new User(this.context, schedulerSockets);
		this.context.sendSetup();
	}

	public void start() {
		this.context.stopSimulation();
	}

	public void stop() {
		this.context.stopSimulation();
	}
}

File: ds-site/src/grid/simulation.ts
import { GridResourceManager } from './grid-resource-manager';
import { GridScheduler } from './grid-scheduler';
import { GridSocket } from './grid-socket';
import { GridUser } from './grid-user';
import { GridWorker } from './grid-worker';

export class Simulation {
    private user: GridUser;
    private schedulers: GridScheduler[] = [];
    private resourceManagers: GridResourceManager[] = [];
    private workers: { [id: string]: GridWorker } = {};

    constructor(
        schedulerCount: number,
        clusterCount: number,
        workerCount: number
    ) {
        this.user = new GridUser();

        const rmSockets: GridSocket[] = [];
        for (let c = 0; c < clusterCount; c++) {
            const rm = new GridResourceManager();
            this.resourceManagers.push(rm);
            rmSockets.push(rm.socket);

            const workerSockets: GridSocket[] = [];
            for (let w = 0; w < workerCount; w++) {
                const worker = new GridWorker();
                this.workers[rm.id] = worker;
                workerSockets.push(worker.socket);
            }

            rm.registerWorkers(workerSockets);
        }

        const schedulerSockets: GridSocket[] = [];
        for (let s = 0; s < schedulerCount; s++) {
            const scheduler = new GridScheduler();
            this.schedulers.push(scheduler);
            schedulerSockets.push(scheduler.socket);
        }

        this.schedulers.forEach(s => {
            s.registerSchedulers(schedulerSockets);
            s.registerResourceManagers(rmSockets);
        });

        this.user.registerSchedulers(schedulerSockets);
        console.log('Simulation initialized...');
    }

    public async start(): Promise<void> {
        this.user.start();
        this.schedulers.forEach(s => s.start());
        this.resourceManagers.forEach(rm => rm.start());

        this.user.test();
    }

    public stop(): void {
        this.user.stop();
        this.schedulers.forEach(s => s.stop());
        this.resourceManagers.forEach(rm => rm.stop());
    }
}
